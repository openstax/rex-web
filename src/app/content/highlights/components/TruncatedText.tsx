import { HTMLElement } from '@openstax/types/lib.dom';
import React from 'react';
import { FormattedMessage } from 'react-intl';
import styled, { css } from 'styled-components/macro';
import { linkStyle } from '../../../components/Typography';
import { textStyle } from '../../../components/Typography/base';
import theme from '../../../theme';
import { cardPadding } from '../constants';

interface Props {
  text: string;
  isActive: boolean;
  className?: string;
  onChange: () => void;
}

// tslint:disable-next-line:variable-name
const Link = styled.span`
  ${textStyle}
  ${linkStyle}
  text-align: left;
  cursor: pointer;
  border: none;
  padding: 0;
  margin: ${cardPadding / 2}rem 0 0 0;
  background: none;
  font-size: 1.2rem;
  line-height: 1.6rem;
`;

// tslint:disable-next-line:variable-name
const NoteText = ({text, isActive, className, onChange }: Props) => {
  const noteTextRef = React.useRef<HTMLElement>(null);
  const [showLink, setShowLink] = React.useState<boolean>(false);

  React.useEffect(() => {
    if (noteTextRef.current && noteTextRef.current.scrollHeight > noteTextRef.current.offsetHeight) {
      setShowLink(true);
    } else {
      setShowLink(false);
    }
    onChange();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isActive]);

  return <React.Fragment>
    <p ref={noteTextRef} className={className}>{text}</p>
    {showLink && <FormattedMessage id='i18n:highlighting:card:show-more'>
      {(msg) => <Link>{msg}</Link>}
    </FormattedMessage>}
  </React.Fragment>;
};

const lineHeight = 1.8;
export default styled(NoteText)`
  ${textStyle}
  overflow: hidden;
  white-space: pre-wrap;
  word-break: break-word;
  overflow-wrap: break-word;
  font-size: 1.4rem;
  line-height: ${lineHeight}rem;
  margin: 0 ${cardPadding / 2}rem 0 0;
  padding: 0;
  ${(props: Props) => props.isActive && css`
    + ${Link} {
      display: none;
    }
  `}
  ${(props: Props) => !props.isActive && css`
    text-overflow: ellipsis;
    overflow: hidden;
    max-height: ${lineHeight * 3}rem;
    display: -webkit-box;
    -webkit-box-orient: vertical;
    -webkit-line-clamp: 3;
  `}

  ${theme.breakpoints.touchDeviceQuery(css`
    padding: ${cardPadding}rem ${cardPadding * 2}rem;
    height: 15.2rem;
    overflow: auto;
  `)}
`;
