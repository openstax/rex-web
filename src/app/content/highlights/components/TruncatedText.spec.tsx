import React from 'react';
import renderer from 'react-test-renderer';
import TestContainer from '../../../../test/TestContainer';
import TruncatedText from './TruncatedText';

describe('TruncatedText', () => {
  it('matches snapshot', () => {
    const component = renderer.create(<TestContainer>
      <TruncatedText text='asdf' isActive={false} onChange={() => null} />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('matches snapshot when focused', () => {
    const component = renderer.create(<TestContainer>
      <TruncatedText text='asdf' isActive={true} onChange={() => null} />
    </TestContainer>);

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('shows link', async() => {
    const createNodeMock = () => ({
      offsetHeight: 50,
      scrollHeight: 100,
    });

    const component = renderer.create(<TestContainer>
      <TruncatedText text='asdf' isActive={true} onChange={() => null} />
    </TestContainer>, {createNodeMock});

    component.update(<TestContainer>
      <TruncatedText text='asdf' isActive={true} onChange={() => null} />
    </TestContainer>);

    expect(() => component.root.findByType('span')).not.toThrow();
  });

  it('calls onChange when state changes', () => {
    const onChange = jest.fn();

    renderer.create(<TestContainer>
      <TruncatedText text='asdf' isActive={false} onChange={onChange} />
    </TestContainer>);

    renderer.create(<TestContainer>
      <TruncatedText text='asdf' isActive={true} onChange={onChange} />
    </TestContainer>);

    expect(onChange).toHaveBeenCalledTimes(1);
  });
});
