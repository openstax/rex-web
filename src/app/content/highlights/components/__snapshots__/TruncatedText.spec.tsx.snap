// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TruncatedText matches snapshot 1`] = `
.c0 {
  color: #424242;
  overflow: hidden;
  white-space: pre-wrap;
  word-break: break-word;
  overflow-wrap: break-word;
  font-size: 1.4rem;
  line-height: 1.8rem;
  margin: 0 0.4rem 0 0;
  padding: 0;
  text-overflow: ellipsis;
  overflow: hidden;
  max-height: 5.4rem;
  display: -webkit-box;
  -webkit-box-orient: vertical;
  -webkit-line-clamp: 3;
}

<p
  className="c0"
>
  asdf
</p>
`;

exports[`TruncatedText matches snapshot when focused 1`] = `
.c0 {
  color: #424242;
  overflow: hidden;
  white-space: pre-wrap;
  word-break: break-word;
  overflow-wrap: break-word;
  font-size: 1.4rem;
  line-height: 1.8rem;
  margin: 0 0.4rem 0 0;
  padding: 0;
}

.c0 + .c1 {
  display: none;
}

<p
  className="c0"
>
  asdf
</p>
`;
